.github/workflows/check.yml

```
@@ -25,34 +25,77 @@ concurrency:
 permissions:
   contents: read
 
+defaults:
+  run:
+    shell: bash
+
 jobs:
+  toolchains:
+    runs-on: ubuntu-latest
+    outputs:
+      toolchains: ${{ steps.toolchains.outputs.toolchains }}
+    steps:
+      - uses: actions/checkout@d632683dd7b4114ad314bca15554477dd762a938 # v4.2.0
+        with:
+          sparse-checkout: Cargo.toml
+      - id: toolchains
+        run: |
+          msrv="$(grep rust-version Cargo.toml | tr -d '"' | cut -f3 -d\ )"
+          echo "toolchains=[\"$msrv\", \"stable\", \"nightly\"]" >> "$GITHUB_OUTPUT"
+
   check:
+    needs: toolchains
     strategy:
       fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        # Keep in sync with Cargo.toml
-        rust-toolchain: [1.76.0, stable, nightly]
+        toolchain: ${{ fromJSON(needs.toolchains.outputs.toolchains) }}
         type: [debug]
+        sanitizer: [null]
         include:
+          # Do one release build on stable, just to check that it works.
           - os: ubuntu-latest
-            rust-toolchain: stable
+            toolchain: stable
             type: release
+          # Add sanitizer builds.
+          - os: ubuntu-latest
+            toolchain: nightly
+            type: debug
+            sanitizer: address
+          - os: ubuntu-latest
+            toolchain: nightly
+            type: debug
+            sanitizer: thread
+          - os: ubuntu-latest
+            toolchain: nightly
+            type: debug
+            sanitizer: leak
+          - os: macos-latest
+            toolchain: nightly
+            type: debug
+            sanitizer: address
+          - os: macos-latest
+            toolchain: nightly
+            type: debug
+            sanitizer: thread
+          # macOS doesn't support the leak checker yet.
+          # - os: macos-latest
+          #   toolchain: nightly
+          #   type: debug
+          #   sanitizer: leak
+
     env:
       BUILD_TYPE: ${{ matrix.type == 'release' && '--release' || '' }}
     runs-on: ${{ matrix.os }}
-    defaults:
-      run:
-        shell: bash
 
     steps:
       - uses: actions/checkout@d632683dd7b4114ad314bca15554477dd762a938 # v4.2.0
 
       - uses: ./.github/actions/rust
         with:
-          version: ${{ matrix.rust-toolchain }}
-          components: ${{ matrix.rust-toolchain == 'stable' && 'llvm-tools-preview' || matrix.rust-toolchain == 'nightly' && 'rust-src' || '' }}
-          tools: ${{ matrix.rust-toolchain == 'stable' && 'cargo-llvm-cov, ' || '' }} cargo-nextest
+          version: ${{ matrix.toolchain }}
+          components: ${{ matrix.toolchain == 'stable' && 'llvm-tools-preview' || matrix.toolchain == 'nightly' && 'rust-src' || '' }}
```

```suggestion
          components: ${{ matrix.toolchain == 'stable' && 'llvm-tools-preview' || matrix.sanitizer != null && 'rust-src' || '' }}
```

Would be more accurate, right?